### PR TITLE
feat: add helm renderer and update helm render integration tests

### PR DESCRIPTION
**Pull request by MarlonGamez**
Fixes: https://github.com/GoogleContainerTools/skaffold/issues/7285

**Description**
This PR adds the helm renderer, which uses `helm template` to render manifests. It also updates and reenables the `TestHelmRender` integration test.

There is also a huge amount of refactoring in this PR to move a lot of the code shared between the helm renderer (`pkg/skaffold/render/renderer/helm`) and helm legacy deployer (`pkg/skaffold/deploy/helm`) into a common package `pkg/skaffold/helm`

### CODE REVIEW

**Comment by aaron-prindle** [comment]
Seems `TestHelmRender` is failing atm:
https://github.com/GoogleContainerTools/skaffold/runs/6614566005?check_suite_focus=true

**Comment by MarlonGamez** [comment]
@aaron-prindle yeah the data there is all correct I believe but the test expects it to be ordered differently. Assuming something changed with helm that's causing this. I'm gonna update the expected data to match, wdyt? Unless there's a flag I'm missing that will allow it to be ordered as we expect

File: hack/boilerplate.sh [hunk 1]
```diff
@@ -18,7 +18,7 @@ set -e
 DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
 BOILERPLATEDIR=$DIR/boilerplate
 
-files=$(python ${BOILERPLATEDIR}/boilerplate.py --rootdir . --boilerplate-dir ${BOILERPLATEDIR})
+files=$(python3 ${BOILERPLATEDIR}/boilerplate.py --rootdir . --boilerplate-dir ${BOILERPLATEDIR})
```
Comment (aaron-prindle) [inline hack/boilerplate.sh:21]: nit: might make sense to split out as this is already a large PR.  Ok to keep though

File: integration/render_test.go [hunk 2]
```diff
@@ -350,60 +326,42 @@ spec:
...
+        ports:
+        - containerPort: 80
+        resources: {}
 ---
-# Source: skaffold-helm/templates/ingress.yaml
 apiVersion: extensions/v1beta1
 kind: Ingress
 metadata:
-  name: skaffold-helm-skaffold-helm
+  annotations: null
```
Comment (aaron-prindle) [inline integration/render_test.go:342]: nit: why annotations: null vs what it was?

File: integration/testdata/helm-render/builds.out.json [hunk 3]
```diff
@@ -0,0 +1 @@
+{"builds":[{"imageName":"gcr.io/k8s-skaffold/skaffold-helm","tag":"gcr.io/k8s-skaffold/skaffold-helm:sha256-nonsenselettersandnumbers"}]}
```
Comment (aaron-prindle) [inline integration/testdata/helm-render/builds.out.json:1]: is this file necessary?

File: integration/testdata/helm-render/skaffold-helm/templates/NOTES.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+Hello Helm
```
Comment (aaron-prindle) [inline integration/testdata/helm-render/skaffold-helm/templates/NOTES.txt:1]: nit: we might not need to include this file

File: pkg/skaffold/deploy/helm/helm.go [hunk 5]
```diff
@@ -68,25 +68,13 @@ import (
 )
 
 var (
-	// versionRegex extracts version from "helm version --client", for instance: "2.14.0-rc.2"
```
Comment (aaron-prindle) [inline pkg/skaffold/deploy/helm/helm.go:71]: nit: perhaps we can add a comment stating that the helm version is no longer relevant/needs-checking in V2 (vs V1) and give a brief explanation

File: pkg/skaffold/helm/util.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+	if slashes == 1 {
+		b = append(b, '\\')
+	}
+	if needsQuotes {
+		return fmt.Sprintf(`"%s"`, string(b))
+	}
+	return string(b)
+}
+
+func ChartSource(r latest.HelmRelease) string {
```
Comment (aaron-prindle) [inline pkg/skaffold/helm/util.go:104]: nit: this was private prior, is this used outside of the helm package currently?  Might make sense to keep private if so (ok to leave as is)

File: integration/testdata/gke_loadbalancer-render/skaffold.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+apiVersion: skaffold/v3alpha1
+kind: Config
+build:
+  artifacts:
+    - image: gke-loadbalancer
+      context: ./
+manifests:
+  helm:
+    releases:
+    # seed test namespace in the release name.
```
Comment (aaron-prindle) [inline integration/testdata/gke_loadbalancer-render/skaffold.yaml:10]: is this actually done here?  In the other yaml there is a namespace var:
 - name: skaffold-helm-{{.TEST_NS}}
 
 Just want to make sure the comment makes sense here

File: pkg/skaffold/helm/bin.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/blang/semver"
+	shell "github.com/kballard/go-shellquote"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/graph"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+var (
+	// VersionRegex extracts version from "helm version --client", for instance: "2.14.0-rc.2"
+	VersionRegex = regexp.MustCompile(`v?(\d[\w.\-]+)`)
```
Comment (aaron-prindle) [inline pkg/skaffold/helm/bin.go:36]: nit: do we still have the same version constraints/checking-required moving from v1 -> v2?  For some reason I thought we maybe didn't need some checks (ok to keep as is though)

**Review by aaron-prindle — APPROVED** [review]
LGTM.  Left some small nits related to if all tests files are necessary and if we actually need all of the old version checking logic (or if for v2 we can stop doing as many checks, thought we had some different constraints w/ render + deploy but not sure.  keeping them the same seems safest TBH)

File: integration/examples/getting-started/build.out [hunk 9]
```diff
@@ -0,0 +1 @@
+{"builds":[{"imageName":"skaffold-example","tag":"skaffold-example:2a73789f489b1fe940be9631e64685142f1fe6371a8b19c62c150a35aae68271"}]}
```
[the remark this answers is not in the document]
Comment (MarlonGamez) [reply integration/examples/getting-started/build.out:1]: thanks for pointing this out

File: integration/testdata/helm-render/skaffold-helm/templates/NOTES.txt (hunk 4 above)
Comment (MarlonGamez) [reply integration/testdata/helm-render/skaffold-helm/templates/NOTES.txt:1]: yea I'll remove this since it's just a leftover from cloning the other example

File: integration/testdata/gke_loadbalancer-render/skaffold.yaml (hunk 7 above)
Comment (MarlonGamez) [reply integration/testdata/gke_loadbalancer-render/skaffold.yaml:10]: nope! This is just a leftover comment from cloning another example. Will remove

File: integration/testdata/helm-render/builds.out.json (hunk 3 above)
Comment (MarlonGamez) [reply integration/testdata/helm-render/builds.out.json:1]: this file is used for the integration test as it gets passed to `skaffold render --build-artifacts=`

File: integration/render_test.go (hunk 2 above)
Comment (MarlonGamez) [reply integration/render_test.go:342]: I'm assuming this has to do with a change in helm 3.x and iterating of lists in templates? But it seems that it provides null as the value when there is an empty list of values

File: pkg/skaffold/helm/bin.go (hunk 8 above)
Comment (MarlonGamez) [reply pkg/skaffold/helm/bin.go:36]: I think we might still need version checking to differentiate between helm 3.0.x and helm 3.1+. I can do some investigation later and figure this out

**Comment by aaron-prindle** [comment]
Not sure how/if this is related to this PR but the failing test seems like a legit failure IIUC:
```
--- FAIL: TestDockerCLICheckCacheFromArgs/multiple_cache-from_images (0.01s)
        cmd_helper.go:208: 
            wanted: docker build . --file C:\Users\RUNNER~1\AppData\Local\Temp\skaffold4[15](https://github.com/GoogleContainerTools/skaffold/runs/6820459087?check_suite_focus=true#step:6:16)5648276\Dockerfile -t tag --cache-from from/image1 --cache-from from/image2
               got: docker build . --file C:\Users\RUNNER~1\AppData\Local\Temp\skaffold4155648276\C:\Users\RUNNER~1\AppData\Local\Temp\skaffold4155648276\Dockerfile -t tag --cache-from from/image1 --cache-from from/image2
    --- FAIL: TestDockerCLICheckCacheFromArgs/cache-from_self_uses_tagged_image (0.02s)
        cmd_helper.go:208: 
            wanted: docker build . --file C:\Users\RUNNER~1\AppData\Local\Temp\skaffold365434937\Dockerfile -t gcr.io/k8s-skaffold/test:tagged --cache-from gcr.io/k8s-skaffold/test:tagged
               got: docker build . --file C:\Users\RUNNER~1\AppData\Local\Temp\skaffold365434937\C:\Users\RUNNER~1\AppData\Local\Temp\skaffold365434937\Dockerfile -t gcr.io/k8s-skaffold/test:tagged --cache-from gcr.io/k8s-skaffold/test:tagged
```
Looks like a path is getting double appended
